show demonstration details in SkillList [SCHOOL-80, SCHOOL-112]

ryon opened the pull request:
## Changes to SkillList in demonstration window

- Nix index # column (seems not-useful? and takes up precious space)
- Consolidate experience info into one column
- Allow wrapping for experience/task names
- Add link to performance task
- Expand and redesign details area to show task-related artifacts and comments
- Add plus icon to "submit evidence" button

## TODO

- Wire up all the new features
  - Task links
  - Attachments
  - Comments
- Clarify sources for artifacts and comments per discussion Aug 31
  - Demonstration artifacts from student added by teacher
  - Demonstration comments from teacher
  - Task artifacts/comments from student
  - Task artifacts/comments from teacher

## Screenshot

<img width="749" alt="Screen Shot 2021-08-31 at 1 06 48 PM" src="https://user-images.githubusercontent.com/1154929/131545994-409d865b-66fa-49a4-822f-15acf49a57e0.png">

ryon committed:
feat: draft design for enhancements to SkillList

ryon committed:
feat: update skill detail dropdown with new groupings

ryon commented:
Updated design with new link groupings and unified comment stream:

**Notes:**

- Update "From [Teacher]" → ~~"Task Attachments, Teacher" (no name)~~  → "Task Attachments" (updated by @themightychris 2021-10-13)
- Update "From [Student]" → ~~"Task Attachments, Student" (no name)~~ → "Task Submissions" (updated by @themightychris 2021-10-13)
- Link groups should hide if empty
- If no links at all, show empty text ("No artifacts or task attachments")
- Empty text "No comments" as well
- Limit 5 comments; if more available, show "See more" link to task

<img width="719" alt="Screen Shot 2021-09-14 at 12 56 54 PM" src="https://user-images.githubusercontent.com/1154929/133301159-7234bb07-ff73-45aa-a846-883a195c9555.png">

themightychris commented on sencha-workspace/packages/slate-cbl/src/view/demonstrations/SkillList.js:
```diff
@@ -116,14 +131,128 @@ Ext.define('Slate.cbl.view.demonstrations.SkillList', {
+                                    </div>
+
+                                    <div class="skill-list-detail-group">
+                                        <h5 class="skill-list-detail-subheading">From [Teacher]</h5>
```
I was just reviewing the code and realized the way the system separates these isn't student vs teacher, but "attachments" vs "submissions" so we should mirror that language here to reflect what we have

BillClinton replied:
These are now labeled "Task Attachments" and "Task Submissions"

themightychris committed:
WIP: add implementation notes

BillClinton committed:
feat: StudentTask accessible from Demonstration

This changes were made in a paired programming session with @themightychris

This will allow the UI to access comments via the include tag
`include=Demonstration.StudentTask.Comments`

BillClinton committed:
WIP: adding data to ui design scafolding

themightychris committed:
feat(submodule): add jarvus-fuzzytime submodule

Used command:

    git submodule add \
        --name "jarvus-fuzzytime" \
        --branch master \
        https://github.com/JarvusInnovations/jarvus-fuzzytime.git \
        sencha-workspace/packages.remote/jarvus-fuzzytime

themightychris committed:
feat(ci): add jarvus-fuzzytime to emergence-site projection

themightychris committed:
feat: add jarvus-fuzzytime to SlateDemonstrationsTeacher app

BillClinton committed:
feat: Add real data to comment display

BillClinton committed:
fix: update jarvus-fuzzytime submodule

BillClinton committed:
fix: add jarvus-fuzzytime package to hololens config

BillClinton committed:
feat: add API data to UI design template

BillClinton committed:
cleanup: remove unused includes

BillClinton committed:
cleanup: remove unnecessary virtual fields

BillClinton committed:
cleanup: remove implementation notes

BillClinton committed:
feat: attachment links to show title or URL

BillClinton committed:
feat: Add fixture data for testing comments

BillClinton committed:
feat: filter out removed submission tasks

BillClinton committed:
fix: remove unnecessary error logging per Chris

BillClinton committed:
feat: Add fixtures to test new features

 - Login as teacher

 - Use Tools menu to navigate to the Competency Dashboard
 - In `Rubric` selection list, select "English Language Arts"
 - in `Students` selection list, select "Example School > Students"
 - resulting URL should be something like:
   ```http://localhost:2191/SlateDemonstrationsTeacher/?apiHost=localhost:2190#ELA/group:students```
 - In grid, expand "Reading Critically".
 - click first cell under "Edmund Ebel" in the row labeled "Choose and apply reading strategies"
 - in pop-up window, click the row with title "The Art of Resistance"
 - Attachments and Comments should be visible

 - Use Tools menu to navigate to the Competency Dashboard
 - In `Rubric` selection list, select "English Language Arts"
 - in `Students` selection list, select "English Language Arts > Slate/Slate > C Band"
 - resulting URL should be something like:
   ```http://localhost:2191/SlateDemonstrationsTeacher/?apiHost=localhost:2190#ELA/section:ELA-001```
 - In grid, expand "Reading Critically".
 - click first cell under "Student Slate" in the row labeled "Choose and apply reading strategies"
 - in pop-up window, click the firt row with `Performance Task` "ELA Task One"
 - Only Comments should be visible and should use full horizontal width of details area

 - Use Tools menu to navigate to the Competency Dashboard
 - In `Rubric` selection list, select "English Language Arts"
 - in `Students` selection list, select "Example School > Students"
 - resulting URL should be something like:
   ```http://localhost:2191/SlateDemonstrationsTeacher/?apiHost=localhost:2190#ELA/group:students```
 - In grid, expand "Reading Critically".
 - click first cell under "Edmund Ebel" in the row labeled "Choose and apply reading strategies"
 - in pop-up window, click the row with title "The Art of Resistance"
 - 3 Task submissions should be displayed.  The 2 removed submissions should not be displayed

themightychris commented:
## TODO

- [x] implement task links
  - should open in new tab
  - should link to either teacher tasks dashboard or student tasks dashboard, based on user's role
  - need to add new deep-linking capability to open task details

BillClinton committed:
fix: update fixture data to reflect all changes

themightychris commented on sencha-workspace/packages/slate-cbl/src/view/demonstrations/SkillList.js:
```diff
@@ -45,27 +55,30 @@ Ext.define('Slate.cbl.view.demonstrations.SkillList', {
-                '<th class="skill-list-header skill-list-demo-context">Exp. Name</th>',
-                '<th class="skill-list-header skill-list-demo-task">Perf. Task</th>',
+                '<th class="skill-list-header skill-list-demo-experience">Experience</th>',
+                // '<th class="skill-list-header skill-list-demo-experience">Exp. Type</th>',
+                // '<th class="skill-list-header skill-list-demo-context">Exp. Name</th>',
```
let's delete all these commented out bits

themightychris commented on sencha-workspace/packages/slate-cbl/src/view/demonstrations/SkillList.js:
```diff
@@ -79,28 +92,27 @@ Ext.define('Slate.cbl.view.demonstrations.SkillList', {
+                            '<div class="skill-list-title">{Demonstration.Context:htmlEncode}</div>',
+                            '<small class="skill-list-subtitle">{Demonstration.ExperienceType:htmlEncode}</small>',
+                        '</td>',
+                        // '<td class="skill-list-demo-data skill-list-demo-type" data-qtip="{Demonstration.ExperienceType:htmlEncode}">{Demonstration.ExperienceType:htmlEncode}</td>',
```
delete these too

BillClinton replied:
removed. 1ac64b7882eb9cff50edce0b2af459ae109078a3

themightychris commented on sencha-workspace/packages/slate-cbl/src/view/demonstrations/SkillList.js:
```diff
@@ -116,16 +128,112 @@ Ext.define('Slate.cbl.view.demonstrations.SkillList', {
+                                          </div>
+                                        </tpl>
+                                        {% var attachments = this.getActiveAttachments(values.Demonstration.StudentTask.Attachments) %}
+                                        <tpl if={attachments.length}>
```
I think it's a fluke here that this does anything significant, I think what you're accomplishing here is forcing XTemplate to process it as a raw javascript snippet by putting a dummy code block around it. The preferred way to force raw expression is to wrap in parens, and it should be quoted either way:

```suggestion
                                        <tpl if="(attachments.length)">
```

BillClinton replied:
I realized I was no longer using the attachments variable defined on the previous line elsewhere, so I removed the variable and updated the conditional to call a template function.  a5cc892025093044422b5ae864cc5f29b8e724d2

BillClinton committed:
fix: add fuzzytime to holo reqs for all apps

BillClinton committed:
fix: remove fuzzytime, included with slate-cbl

BillClinton committed:
fix: add blank target for artifact URLs

themightychris commented on sencha-workspace/packages/slate-cbl/src/view/demonstrations/SkillList.js:
```diff
@@ -79,28 +92,27 @@ Ext.define('Slate.cbl.view.demonstrations.SkillList', {
+                            '<a class="skill-list-linkable" href="#">',
+                                '<i class="linkable-icon fa fa-external-link-square"></i>',
+                                '<div class="linkable-content">',
+                                    '{Demonstration.PerformanceType:htmlEncode}',
```
this should be `Demonstration.StudentTask.Task.Title` and this whole link should be wrapped in `<tpl if="Demonstration.StudentTask">`

BillClinton replied:
These changes have been made

BillClinton committed:
feat: add task link

BillClinton committed:
fix: correct attribute used in text in link text

BillClinton committed:
refactor: create template fnction for task link

BillClinton committed:
feat: enable direct linking to student task popup

BillClinton committed:
refactor: remove commented lines from template

BillClinton committed:
fix: correct submission attachment conditional

BillClinton committed:
fix: add conditional around task link block

BillClinton committed:
fix: updated conditional

ryon committed:
fix: break long strings in comments/attachments

themightychris commented on php-classes/Slate/CBL/Demonstrations/Demonstration.php:
```diff
@@ -53,6 +54,12 @@ class Demonstration extends \VersionedRecord
+          'type' => 'one-one',
+          'class' => StudentTask::class,
+          'local' => 'ID',
+          'foreign' => 'DemonstrationID'
```
@BillClinton one tweak to add here: there is no unique constraint on on the `DemonstrationID` column in the `StudentTask` table, so theoretically there can be multiple `StudentTask` records pointing to the same `Demonstration`. 

We should either add a unique constraint to the `DemonstrationID` column, or adjust this relationship config to at least make it deterministic in the case of multiple matches existing

```suggestion
          'foreign' => 'DemonstrationID',
          'order' => ['ID' => 'DESC']
```

would suffice to at least make sure the newest match is always returned

themightychris replied:
@BillClinton I went ahead and added the new unique constraint in 02e2cd07fb2c8fe71d90064ca0f590d0617701d6

themightychris replied:
so just read this for FYI but then mark resolved, no need to add the order but FYI that you can do that

themightychris commented on sencha-workspace/packages/slate-cbl/src/view/demonstrations/SkillList.js:
```diff
@@ -116,16 +125,114 @@ Ext.define('Slate.cbl.view.demonstrations.SkillList', {
+                                                </ul>
+                                            </div>
+                                        </tpl>
+                                        <tpl if="Demonstration.StudentTask.Task.Attachments && Demonstration.StudentTask.Task.Attachments.length">
```
@BillClinton Task-level attachments use the same model and can also have `Status=removed` so we should probably use the same filtering here as we do below (and the same `Status != "removed` filter within the loop)

BillClinton replied:
resolved here: f858d846999940a6d9b5c02b1e806819b8e6dabb

themightychris commented on sencha-workspace/packages/slate-cbl/src/view/demonstrations/SkillList.js:
```diff
@@ -116,16 +125,114 @@ Ext.define('Slate.cbl.view.demonstrations.SkillList', {
+                                                <ul class="skill-list-links">
+                                                    <tpl for="Demonstration.StudentTask.Task.Attachments">
+                                                        <li class="skill-list-link-item">
+                                                            <a href="{URL}" target="_blank" class="skill-list-link">
```
```suggestion
                                                            <a href="{URL:htmlEncode}" target="_blank" class="skill-list-link">
```

BillClinton replied:
resolved 1ffe9c6d165268ee08b8824413a0a8aa06eace14

themightychris commented on sencha-workspace/packages/slate-cbl/src/view/demonstrations/SkillList.js:
```diff
@@ -116,16 +125,114 @@ Ext.define('Slate.cbl.view.demonstrations.SkillList', {
+                                                    <tpl for="Demonstration.StudentTask.Attachments">
+                                                    <tpl if="Status != 'removed'">
+                                                            <li class="skill-list-link-item">
+                                                                <a href="{URL}" target="_blank" class="skill-list-link">
```
```suggestion
                                                                <a href="{URL:htmlEncode}" target="_blank" class="skill-list-link">
```

BillClinton replied:
resolved 1ffe9c6d165268ee08b8824413a0a8aa06eace14

themightychris committed:
feat(db): add unique constraint to StudentTask.DemonstrationID

themightychris commented on sencha-workspace/packages/slate-cbl/src/view/demonstrations/SkillList.js:
```diff
@@ -116,16 +125,114 @@ Ext.define('Slate.cbl.view.demonstrations.SkillList', {
+                                                                {[Ext.util.Format.fuzzyTime(createDate,true)]}
+                                                            </time>
+                                                        </div>
+                                                        <div class="skill-list-comment-text">{Message}</div>
```
```suggestion
                                                        <div class="skill-list-comment-text">{Message:htmlEncode}</div>
```

BillClinton replied:
resolved 1ffe9c6d165268ee08b8824413a0a8aa06eace14

themightychris commented on sencha-workspace/packages/slate-cbl/src/view/demonstrations/SkillList.js:
```diff
@@ -116,16 +125,114 @@ Ext.define('Slate.cbl.view.demonstrations.SkillList', {
+                                                            <li class="skill-list-link-item">
+                                                                <a href="{URL}" target="_blank" class="skill-list-link">
+                                                                    <i class="fa fa-link skill-list-link-icon"></i>
+                                                                    <div class="skill-list-link-label">{[this.linkText(values)]}</div>
```
```suggestion
                                                                    <div class="skill-list-link-label">{[fm.htmlEncode(this.linkText(values))]}</div>
```

I _think_ this will work in this context, xtemplate javascript has access to an `fm` alias to `Ext.util.Format`

BillClinton replied:
resolved 1ffe9c6d165268ee08b8824413a0a8aa06eace14

themightychris commented on sencha-workspace/packages/slate-cbl/src/view/demonstrations/SkillList.js:
```diff
@@ -116,16 +125,114 @@ Ext.define('Slate.cbl.view.demonstrations.SkillList', {
+                                                        <li class="skill-list-link-item">
+                                                            <a href="{URL}" target="_blank" class="skill-list-link">
+                                                                <i class="fa fa-link skill-list-link-icon"></i>
+                                                                <div class="skill-list-link-label">{[this.linkText(values)]}</div>
```
```suggestion
                                                                <div class="skill-list-link-label">{[fm.htmlEncode(this.linkText(values))]}</div>
```

BillClinton replied:
resolved 1ffe9c6d165268ee08b8824413a0a8aa06eace14

BillClinton committed:
fix: filter task attachments by removed flag

BillClinton committed:
fix: correct indent

BillClinton committed:
fix: escape HTML on displayed fields

themightychris commented:
I just rebased and force-pushed this to get on top of latest fixed tests...i had to resolve some conflicts in the fixtures, we need to re-test the PR environment to make sure the added fixture data is still working correctly

themightychris reviewed (approved):
fantastic!
